### DISCUSSION TITLE
What is the real reason to tuya magic spell ?

### DISCUSSION THREAD

**Discussion by w35l3y** [zigpy/zha-device-handlers#2032] [General]
In other words, what doesn't work without the tuya magic spell ?
Does it enable some hidden functionality/cluster ?

**Comment by MattWestb** [comment]
We dont knowing so mush what is doing then its not documented thing and the community was finding then debugging tuya TS004F remotes not using clusters on not existing endpoints.
Its 2 classes its working. Firsst is if the device is changing endpoints 7 cluster it must being sent very early then joining the device then the device still is init mode. tuya is configuring the device and then sending one leave with rejoin so the device is leaving and coming back and is in init mode for some mili seconds and can sending the attributes to the device and the device is activating the new endpoints and reporting it the coordinator then it reading it.
The second version is only "software" = no change or endpoints / cluster config it can being sent every time and is working. One example is tuya / LIDL power strip with new firmware, its have 4 outlets and before getting the tuya magic change status of all outputs as it was only one switch for all. After getting it its working like 4 separate outlets so swishing output 2 with command is only switching output 2 in the hardware.

As i was saying we dont knowing what its doing we only knowing tuya is using it for all there device in there system then looking in sniffs from tuya paring and its fixing some strange implanting tuya have doing.

I think we can see it as one working mode of the device, "normal" Zigbee mode and tuya "extended" mode.
Or that tuya have doing the firmware bad and need making things working by running scripts in the device that is triggered if reading this attributes and fixing some bad things in the code on the fly.

**Reply by w35l3y** [reply]
Do you know any source about magic spell so I could replicate it at another platform ? (SmartThings drivers)

**Reply by w35l3y** [reply]
https://github.com/w35l3y/EdgeDrivers/blob/0db10b0243786e73497eabaae644bce2237e6b9c/personal-tuya-devices/src/lifecycles.lua#L78-L86

Just for historical purposes.

I received a report today that this code worked for an user of thermostat ( _TZE200_4hbx5cvx )

Thanks for clarifying what `tuya magic spell` is.

**Reply by gilrim** [reply]
Has this fix been contributed back to the repositories? I've got this device and would like to get it working :)

**Reply by w35l3y** [reply]
> Has this fix been contributed back to the repositories?

Which ones ?
https://github.com/w35l3y/EdgeDrivers/issues/2 ?

**Reply by gilrim** [reply]
Yes indeed! Didn't find that when looking yesterday: will it work with ha/zha as a quirk directly?

**Reply by w35l3y** [reply]
No. My case is for another platform.
I asked here because @MattWestb is an expert at tuya magic spell.
I studied this repo to make TMS work in my case.

**Comment by w35l3y** [comment]
In my case, the device is 6-gangs switch that uses 0xEF00 cluster.
First I tried to add it using Home Assistant, but it is not easy at all to debug a single device. I only have 1 dongle with a lot devices paired with. I don't know how to sniff that single device.
So I decided to pair that device with the new SmartThings Driver system.
It allows me to filter the logs for that single device, definitelly much easier to debug.
I thought I needed it because it doesn't report 0x0006 cluster, but it works individually when commands are sent directly to 0xEF00 cluster.

Currently, my problem is that the device reports offline after awhile.
It turns online again when we press any gang that sends event to coordinator.
That is the problem I currently am trying to fix.

I know how to fix it when there is 0x0006 cluster.
I used to send read_attribute command to read 0x0000 attribute.

But without that cluster, I don't know how to fix it.
0xEF00 cluster doesn't seem to have any attribute so I could force reading it.

That is why I thought: maybe I should try that magic spell, so 0x0006 cluster might show up.

That is why I am trying to understand magic spell to know how and when to use it properly.

**Reply by MattWestb** [reply]
Is it one TS601 device (ala tuya MCU) ?

**Reply by w35l3y** [reply]
https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/tuya/ts0601_switch.py#L574
This is the quirk I developed.

**Reply by MattWestb** [reply]
All TS601 device is tuya MCU device = they is having one MCU that is doing the device things and sending commands to the Zigbee module over one com port and its tunneling the commands to tuya cloud.
ZHA is using quirks for decoding the commands and making one new device with endpoints and attributes.
tuya magic can helping the Zigbee module working better but all with all TS601 is is making quirks that is adding  the futures to ZHa that the device (MCU) is having.
So if your TS601 device is missing function then the quirk need being updated and tuya magick cant helping  fixing the missing functions.

**Reply by gilrim** [reply]
I'm trying to figure out what I can do to fix the integration for tuya thermostat linked here: https://github.com/zigpy/zha-device-handlers/issues/1849
I've tried adding the device code to the electric heating quirk, but that made no difference for me: https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/tuya/ts0601_electric_heating.py

How / where should I add the bits about magic spells?

**Reply by MattWestb** [reply]
You need deleting the device in ZHA and waiting one minute and then joining it new for getting the spell being casted then its in INIT mode. Also deleting and restarting is better but taking longer.
